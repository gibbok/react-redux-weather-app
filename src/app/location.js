// @flow
import {LocationType} from '../types'  // eslint-disable-line no-unused-vars

const location = ():LocationType => {
  return {
    locationId: 5128581, // nyc usa
    geo: [-74.006, 40.7143],
    isFetching: false
  }
}

export default location
