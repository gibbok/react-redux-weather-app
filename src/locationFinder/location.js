const Location = ({onClick, name, country}) => (
  <li onClick={onClick}>
    {name} {country}
  </li>
)

export default Location
