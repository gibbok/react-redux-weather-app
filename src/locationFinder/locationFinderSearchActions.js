import 'isomorphic-fetch'

const getFindLocations = q => ({
    type: 'GET_FIND',
    payload: new Promise(resolve => {
        setTimeout(() => fetch(`http://api.openweathermap.org/data/2.5/find?q=${q}&APPID=9f585babc59bf435c42319b4e9b69766`).then(response => {
            resolve(response.json());
        }), 1000);
    })
});

const setFinderSearchValue = (value) => ({
    type: 'SET_FIND_VALUE',
    payload: new Promise(resolve => {
            resolve(value);
    })
});

export { getFindLocations }