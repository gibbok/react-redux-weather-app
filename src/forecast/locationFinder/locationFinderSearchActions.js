// @flow
import {ActionType} from '../../types' // eslint-disable-line no-unused-vars

import 'isomorphic-fetch'
import * as api from '../../app/api'
import * as types from './locationFinderActionTypes'
import { getWeather } from '../../forecast/weather/weatherActions'
import { getForecastDaily } from '../../forecast/forecastDaily/forecastDailyActions'
import { getForecastHourly } from '../../forecast/forecastHourly/forecastHourlyActions'

const getLocations = (query:string):ActionType => {
  const value = query.trim()

  if (!value.length) {
    return ({
      type: null,
      payload: null
    })
  }

  return ({
    type: types.GET_LOCATIONS,
    payload: new Promise((resolve, reject) => {
      fetch(api.find(query)).then(response => {
        resolve(response.json())
      })
    })
  })
}
const setSearchValue = (value:string):ActionType => ({
  type: types.SET_SEARCHVALUE,
  payload: value
})

const setLocation = (id:number):ActionType => {
  return dispatch => {
    dispatch(getWeather(id))
    dispatch(getForecastDaily(id))
    dispatch(getForecastHourly(id))
  }
}

export { getLocations, setSearchValue, setLocation }
