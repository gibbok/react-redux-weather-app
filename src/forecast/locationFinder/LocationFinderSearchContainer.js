// @flow
import { connect } from 'react-redux'
import LocationFinderSearch from './LocationFinderSearch'
import { getLocations, setSearchValue, setIsFinderOpen } from './locationFinderSearchActions'
import type { Dispatch } from 'redux' // eslint-disable-line no-unused-vars

const getInputValue = state => {
  return state.locationFinderReducer.locationFinder.ui.inputValue
}

const mapStateToProps = (state, ownProps) => {
  return {
    inputValue: getInputValue(state)
  }
}

/* istanbul ignore next */
const mapDispatchToProps = (dispatch:Dispatch<*>, ownProps) => {
  return {
    onSearchClick: (e, inputValue) => {
      e.preventDefault()
      dispatch(setIsFinderOpen(true))
      dispatch(getLocations(inputValue))
    },
    onLocationChange: e => {
      e.preventDefault()
      const value = e.target.value
      dispatch(setSearchValue(value))
    },
    onLocationKeyDown: e => {
      if (e.key === 'Enter') {
        e.preventDefault()
        const value = e.target.value
        dispatch(setSearchValue(value))
        dispatch(getLocations(value))
      }
    }
  }
}

const LocationFinderSearchContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(LocationFinderSearch)

export default LocationFinderSearchContainer
