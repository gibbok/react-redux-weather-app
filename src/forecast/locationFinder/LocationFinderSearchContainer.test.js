import React from 'react'
import { createMockStore, createMockDispatch } from 'redux-test-utils'
import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import configureStore from 'redux-mock-store'
import LocationFinderSearchContainer from './LocationFinderSearchContainer'
import { getLocations, setLocation, setSearchValue } from './locationFinderSearchActions'
import location from '../../app/location'

describe('<LocationFinderSearchContainer />', () => {
  it('should render', () => {
    const initialState = {
      locationFinderReducer: {
        locationFinder: {
          data: [
          ],
          app: {
            isFetching: false
          },
          ui: {
            inputValue: ''
          }
        }
      }
    }
    const mockStore = configureStore()
    const store = mockStore(initialState)
    const wrapper = shallow(<LocationFinderSearchContainer store={store} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should dispatch action getLocations', () => {
    const state = 'state'
    const store = createMockStore(state)
    const action = getLocations(location().locationId)

    store.dispatch(action)
    expect(store.getAction(action.type)).toEqual(action)
    expect(store.getActions()).toEqual([action])
    expect(store.isActionDispatched(action)).toBe(true)
    expect(store.isActionTypeDispatched(action.type)).toBe(true)
    expect(store.getState()).toBe(state)

    const dispatchMock = createMockDispatch()
    dispatchMock.dispatch(action)
    expect(dispatchMock.getAction(action.type)).toEqual(action)
    expect(dispatchMock.getActions()).toEqual([action])
    expect(dispatchMock.isActionDispatched(action)).toBe(true)
    expect(dispatchMock.isActionTypeDispatched(action.type)).toBe(true)
  })

  it('should dispatch action setLocation', () => {
    const state = 'state'
    const store = createMockStore(state)
    const action = setLocation(location().locationId)

    store.dispatch(action)
    expect(store.getAction(action.type)).toEqual(action)
    expect(store.getActions()).toEqual([action])
    expect(store.isActionDispatched(action)).toBe(true)
    expect(store.isActionTypeDispatched(action.type)).toBe(true)
    expect(store.getState()).toBe(state)

    const dispatchMock = createMockDispatch()
    dispatchMock.dispatch(action)
    expect(dispatchMock.getAction(action.type)).toEqual(action)
    expect(dispatchMock.getActions()).toEqual([action])
    expect(dispatchMock.isActionDispatched(action)).toBe(true)
    expect(dispatchMock.isActionTypeDispatched(action.type)).toBe(true)
  })

  it('should dispatch action setSearchValue', () => {
    const state = 'state'
    const store = createMockStore(state)
    const action = setSearchValue('brno')

    store.dispatch(action)
    expect(store.getAction(action.type)).toEqual(action)
    expect(store.getActions()).toEqual([action])
    expect(store.isActionDispatched(action)).toBe(true)
    expect(store.isActionTypeDispatched(action.type)).toBe(true)
    expect(store.getState()).toBe(state)

    const dispatchMock = createMockDispatch()
    dispatchMock.dispatch(action)
    expect(dispatchMock.getAction(action.type)).toEqual(action)
    expect(dispatchMock.getActions()).toEqual([action])
    expect(dispatchMock.isActionDispatched(action)).toBe(true)
    expect(dispatchMock.isActionTypeDispatched(action.type)).toBe(true)
  })
})
