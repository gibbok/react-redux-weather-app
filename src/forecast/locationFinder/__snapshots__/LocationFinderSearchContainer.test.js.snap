// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationFinderSearchContainer /> should render 1`] = `
<LocationFinderSearch
  inputValue=""
  onLocationChange={[Function]}
  onLocationKeyPress={[Function]}
  onSearchClick={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "notify": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": null,
    }
  }
/>
`;
